### DISCUSSION TITLE
`affectsConfiguration()` returns `true` when a project is removed in a multi-workspace instance

### DISCUSSION THREAD

**Discussion by OrkunTokdemir** [microsoft/vscode-discussions#2408] [Extension Development QnA]
`affectsConfiguration("my_config", my_workspace_folder)` returns `true` when I remove `my_workspace_folder` from the multi-workspace instance. This leads to bugs on our end. I would expect it to be not triggered for those situations. Is there any workaround for this problem?

I have a similar configuration in my workspace `setting.json`
```json
"my_config.myval": true
```

**Comment by isidorn** [comment]
Well that sounds like the right behaviour to me. Why is this unexpected?

**Reply by OrkunTokdemir** [reply]
I don't understand the relationship between removing the workspace folder and sending a configuration change notification. When our implementation gets the configuration change notification, it starts a thread as if the user changed their configuration but the user doesn't change the configuration, just removes the workspace folder.  In the current vscode API implementation, extension developers should figure out whether it is a real notification change or removing the workspace folder which is not easy. I would expect the below logic 

```
configuration change -> notify about only configuration change
removed folder -> notify about only folder removal
```
instead of 

```
configuration change -> notify about only configuration change
removed folder -> notify about configuration change & notify about folder removal
```

**Reply by isidorn** [reply]
@sandy081 might be able to help

**Reply by sandy081** [reply]
Because removing a folder has a side affect of removing the workspace configuration file from that folder, hence there is a configuration change event

**Reply by OrkunTokdemir** [reply]
That would be right if the scope parameter is not passed, but in my example, I explicitly passed the workspace folder scope. What I understand from what you said is that the vscode implementation sends a notification like just the workspace configuration file is removed by the user explicitly and then it sends another notification for the workspace removal. Can you tell me a use case where this design is useful? Because it leads clients to implement workaround solutions to check why `affectsConfiguration()` returns true.

**Reply by sandy081** [reply]
Can you please check this issue and let me know if you are hitting the same - https://github.com/microsoft/vscode/issues/66246 - If so I have mentioned a solution how to tackle this specific case

**Reply by olguzzar** [reply]
Hi @OrkunTokdemir, did the solution mentioned in the above issue work for you?

**Reply by OrkunTokdemir** [reply]
The workaround shared in https://github.com/microsoft/vscode/issues/66246#issuecomment-454441345 doesn't fix my problem. To be more specific, when a configuration is changed, our extension restarts our language server with the new parameters. When `affectsConfiguration()` is triggered before `onDidChangeWorkspaceFolders()` by removing, we start a dangling language server. Our workaround would be either stopping the dangling language server or aborting starting it when `onDidChangeWorkspaceFolders()`  is triggered but it is a dirty workaround which I avoided from the beginning.

**Reply by sandy081** [reply]
> To be more specific, when a configuration is changed, our extension restarts our language server with the new parameters.

Can you please explain this more? Do you restart on any configuration change or for a specific change?

**Reply by OrkunTokdemir** [reply]
We restart it using a specific language server configuration change. For example, users can provide some import paths for the language server. When an import path is either added or removed, the language server is restarted with the new import paths.

**Reply by sandy081** [reply]
So why not checking if that particular configuration has changed or not whenever there is a config change event. Restart only when that specific config has changed.

**Reply by OrkunTokdemir** [reply]
We already check for a particular configuration change. 

Let me explain in more detail.  

`affectsConfiguration(<some_config>, my_workspace_folder)` -> Checks the specific configuration and restarts the language server with new parameters
`onDidChangeWorkspaceFolders()` -> If a workspace is removed, the extension stops the language server.

When a workspace is removed, `affectsConfiguration()` returns true, and the extension restarts the language server. Then, `onDidChangeWorkspaceFolders()` is triggered, and the workspace is removed, but it cannot stop the language server because it is being initialized. In the end, we don't have the workspace but a language server working for the removed workspace, which becomes a kind of dangling language server.

I think there is a bug in `affectsConfiguration()` for the case when the workspace parameter is passed and the workspace is removed. It should return false but returns true, which misleads the execution on the extension side. There is no way to understand whether the configuration is removed by the user or by the workspace removal.

**Reply by sandy081** [reply]
> I think there is a bug in affectsConfiguration() for the case when the workspace parameter is passed and the workspace is removed. It should return false but returns true, which misleads the execution on the extension side. There is no way to understand whether the configuration is removed by the user or by the workspace removal.

This is interesting. It would be great if you can draft a simple repro surfacing this issue so that it will be easy for us to investigate.
